drivers: spi: improve initialization of multiple cs gpios

bbilas opened the pull request:
Initialize all defined cs gpios and set them as inactive during driver initialization to be sure that before first communication over SPI the cs gpio is in the valid state. It also fixes the case when we have multiple devices connected to the one SPI interface and the initial state of cs gpios after MCU reset is floating and might be low, so that prevents us from communicating between particular devices. 

PR has been tested on STM32F4 with 15 TMC5160 stepper motor controllers connected via SPI.

That's an updated version of https://github.com/zephyrproject-rtos/zephyr/pull/36982

bbilas committed:
drivers: spi: spi_context: improve support of multiple cs gpios

Add extra cs_gpios and num_cs_gpios members into
spi_context structure that will be used to
initialize all defined cs gpios during the driver
initialization using SPI_CONTEXT_CS_GPIOS_INITIALIZE macro.
While at it add a new spi_context_cs_configure_all
function that allows the user to configure
all available cs gpios in inactive mode.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

gmarull commented on drivers/spi/spi_context.h:
```diff
@@ -57,6 +59,20 @@ struct spi_context {
 
+#define SPI_CONTEXT_CS_GPIO_SPEC_ELEM(node_id, prop, idx)				\
+	{ 										\
+		.gpio_dev = DEVICE_DT_GET(DT_GPIO_CTLR_BY_IDX(node_id, prop, idx)),	\
+		.gpio_pin = DT_GPIO_PIN_BY_IDX(node_id, prop, idx),			\
+		.gpio_dt_flags = DT_GPIO_FLAGS_BY_IDX(node_id, prop, idx),		\
```
`gpio_dt_spec` (and associated APIs) can be used instead, will make things easier

bbilas replied:
I could have used this structure but It's not compatible with spi_cs_control structure that I'm using here.

gmarull replied:
then, wouldn't it be better to cleanup that part, too?

mniestroj replied:
@gmarull It would. But I see that there are plenty of drivers to be modified to be `gpio_dt_spec` compatible, so I think this is a topic for another PR. Besides, I would probably first convert drivers to use `spi_dt_spec`, so after that the conversion to `gpio_dt_spec` would be probably just a change of single macro (`SPI_CS_CONTROL_PTR_DT`).

JordanYates replied:
FYI I have started converting drivers to `spi_dt_spec` in #37358 and have commits to convert the spi_cs_control to `gpio_dt_spec` here: https://github.com/csiro-wsn/zephyr/commits/210801_spi_cs

tbursztyka replied:
perhaps it would be better to get spi_cs_control to gpio_dt_spec first then, and afterwards this patch

bbilas replied:
@JordanYates When those changes will be ready? I can't see any MR with them.

mniestroj commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -818,6 +818,24 @@ static const struct spi_driver_api api_funcs = {
 	.release = spi_stm32_release,
 };
 
+static int spi_stm32_cs_gpios_configure(const struct device *dev)
+{
+	int err;
+	const struct spi_cs_control *cs_gpio;
+	struct spi_stm32_data *data = dev->data;
+
+	for (cs_gpio = data->ctx.cs_gpios;
+	     cs_gpio < &data->ctx.cs_gpios[data->ctx.num_cs_gpios];
+	     cs_gpio++) {
+		err = spi_cs_configure(cs_gpio);
+		if (err) {
+			return err;
+		}
+	}
+
+	return 0;
+}
+
```
There is nothing stm32 specific in this function, so I would rather have one generic function that already iterates through all CS gpios.

bbilas replied:
I'm aware of that but I wasn't sure if I should create another function like e.g spi_cs_configure_all that will call spi_cs_configure or create the separate functions for each arch SPI driver and call them in the init function. Now I see that a generic approach would be more convenient so I'm gonna fix that.

bbilas committed:
drivers: spi_ll_stm32: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas commented:
@ABOSTM @FRASTM @erwango gentle ping

erwango commented:
Set as DNM waiting for @tbursztyka approval

gmarull commented on drivers/spi/spi_context.h:
```diff
@@ -183,17 +203,44 @@ gpio_dt_flags_t spi_context_cs_active_level(struct spi_context *ctx)
+			   GPIO_OUTPUT_INACTIVE);
+}
+
+static int __attribute__((unused)) spi_context_cs_configure_all(struct spi_context *ctx)
```
Function defined in a header should be `static inline`

bbilas replied:
Fixed.

gmarull commented on drivers/spi/spi_context.h:
```diff
@@ -57,6 +59,24 @@ struct spi_context {
 #define SPI_CONTEXT_INIT_SYNC(_data, _ctx_name)				\
 	._ctx_name.sync = Z_SEM_INITIALIZER(_data._ctx_name.sync, 0, 1)
 
+#define SPI_CONTEXT_CS_GPIO_SPEC_ELEM(_node_id, _prop, _idx)				\
+	{										\
+		.gpio_dev = DEVICE_DT_GET(DT_GPIO_CTLR_BY_IDX(_node_id, _prop, _idx)),	\
+		.gpio_pin = DT_GPIO_PIN_BY_IDX(_node_id, _prop, _idx),			\
+		.gpio_dt_flags = DT_GPIO_FLAGS_BY_IDX(_node_id, _prop, _idx),		\
+	},
```
this is duplicating `GPIO_DT_SPEC_GET_BY_IDX`, but with different field names, why not modify SPI to use gpio_dt_spec?

bbilas replied:
spi_cs_control struct does not use gpio_dt_spec so I've decided to not touch this and use it as it's and convert that later when it will be done. Take a look at the above conversation.

gmarull replied:
I think it's better to _modernize_ SPI other than adding duplicate code on top of it.

bbilas replied:
Sure, I just didn't want to duplicate the work that @JordanYates has done here https://github.com/csiro-wsn/zephyr/commits/210801_spi_cs

bbilas replied:
I'll change that when https://github.com/zephyrproject-rtos/zephyr/pull/39116 will be merged.

bbilas replied:
Fixed.

gmarull commented on drivers/spi/spi_context.h:
```diff
@@ -183,17 +203,44 @@ gpio_dt_flags_t spi_context_cs_active_level(struct spi_context *ctx)
 
+static inline int spi_cs_configure(const struct spi_cs_control *cs_gpio)
+{
+	return gpio_pin_configure(cs_gpio->gpio_dev,
+			   cs_gpio->gpio_pin,
+			   cs_gpio->gpio_dt_flags |
+			   GPIO_OUTPUT_INACTIVE);
```
why not modify SPI to use `gpio_dt_spec`?

bbilas replied:
as above

bbilas replied:
Fixed.

gmarull commented on drivers/spi/spi_context.h:
```diff
@@ -183,17 +203,44 @@ gpio_dt_flags_t spi_context_cs_active_level(struct spi_context *ctx)
+	     cs_gpio < &ctx->cs_gpios[ctx->num_cs_gpios];
+	     cs_gpio++) {
+		err = spi_cs_configure(cs_gpio);
+		if (err) {
```
`err < 0` (essentially boolean expression)

bbilas replied:
Fixed.

gmarull commented on drivers/spi/spi_context.h:
```diff
@@ -183,17 +203,44 @@ gpio_dt_flags_t spi_context_cs_active_level(struct spi_context *ctx)
-				   ctx->config->cs->gpio_pin,
-				   ctx->config->cs->gpio_dt_flags |
-				   GPIO_OUTPUT_INACTIVE);
+		ret = spi_cs_configure(ctx->config->cs);
+		if (ret < 0) {
+			LOG_ERR("Failed to configure 'cs' gpio: %d", ret);
+		}
```
`spi_context_cs_configure` should be modified to return error code, this will hide the problem, e.g. if no logging is enabled.

bbilas replied:
I saw it but I thought that someone had some serious reason to ignore the return code from `gpio_pin_configure` function so I haven't changed that. I will fix it.

bbilas replied:
Fixed.

JordanYates commented on drivers/spi/spi_context.h:
```diff
@@ -183,6 +196,28 @@ gpio_dt_flags_t spi_context_cs_active_level(struct spi_context *ctx)
 	return GPIO_ACTIVE_LOW;
 }
 
+static inline int spi_cs_configure(const struct spi_cs_control *cs_gpio)
```
The naming of these functions is misleading to me, they are all `_configure` but they only configure in one direction, unlike the GPIO functions for example. Perhaps `initialize`?

JordanYates commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -846,6 +846,11 @@ static int spi_stm32_init(const struct device *dev)
 		return err;
 	}
 
+	err = spi_context_cs_configure_all(&data->ctx);
```
If this logic makes sense generically, the rest of the SPI drivers should also be updated to do this.
If it does not make sense generally, the new functions probably shouldn't be in the global headers.
I don't think we want to introduce more family specific behaviors if we can help it (SPI CS lines are setup on STM, but not any others).

bbilas replied:
The aim is to provide the same support for all supported drivers but currently I have only one device with stm32f4 that can be used for testing. Next week, I'll get some devboards with nRF52 chips so I'll add these changes for them too.

bbilas committed:
drivers: spi_b91: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_cc13xx_cc26xx: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_dw: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_esp32_spim: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_gecko: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_mcux_dspi: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_mcux_flexcomm: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_mcux_lpspi: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_nrfx_spi: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_nrfx_spim: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_oc_simple: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_psoc6: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_rv32m1_lpspi: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_sam: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_sam0: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_sifive: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_xec_qmspi: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi_xlnx_axi_quadspi: initialize all cs gpios during init

In case when we have multiple devices connected to the
one SPI interface the initial state of CS gpios after
MCU reset is floating and it might be low that prevents us from
communicating between particular devices. Fix that by
initializing all provided cs gpios and setting them as inactive.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

gmarull reviewed (commented):
I assume all `spi_context_cs_configure` could now be deleted from drivers, right?

Since `spi_context_cs_configure_all` is called during init, it should also check for GPIO device readiness, and remove that check from `spi_ready()`. 

I wonder if there should also be another helper that calls `gpio_configure` with `GPIO_DISCONNECTED` (for PM purposes), but that's probably for another PR.

Just some questions after getting some more context.

bbilas commented:
> I assume all `spi_context_cs_configure` could now be deleted from drivers, right?
> 
> Since `spi_context_cs_configure_all` is called during init, it should also check for GPIO device readiness, and remove that check from `spi_ready()`.
> 
I think so, we can do something like that
<pre>
static inline int spi_context_cs_configure_all(struct spi_context *ctx)
{
	int ret;
	const struct gpio_dt_spec *cs_gpio;

	for (cs_gpio = ctx->cs_gpios; cs_gpio < &ctx->cs_gpios[ctx->num_cs_gpios]; cs_gpio++) {
		if (!device_is_ready(cs_gpio->port) {
	                LOG_ERR("GPIO port %s is not ready", cs_gpio->port->name);
	                return -ENODEV;
		}

		/* Validate CS active levels are equivalent */
		__ASSERT(spi_context_cs_active_level(ctx) ==
			 (cs_gpio->gpio.dt_flags & GPIO_ACTIVE_LOW),
			 "Devicetree and spi_context CS levels are not equal");

		ret = gpio_pin_configure_dt(cs_gpio, GPIO_OUTPUT_INACTIVE);
		if (ret < 0) {
			return ret;
		}
	}

	return 0;
}
</pre>
> I wonder if there should also be another helper that calls `gpio_configure` with `GPIO_DISCONNECTED` (for PM purposes), but that's probably for another PR.
> 
> Just some questions after getting some more context.

I'm not familiar with power management in zephyr so it's hard for me to say something.

bbilas committed:
drivers: spi: remove spi_context_cs_configure function

Since cs gpios are initialized during driver initialization
remove spi_context_cs_configure that is not longer need.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

bbilas committed:
drivers: spi: use dedicated init priority

A couple of SPI drivers use CONFIG_KERNEL_INIT_PRIORITY_DEVICE
as init priority for driver initialization. Let's change
it to the dedicated CONFIG_SPI_INIT_PRIORITY to make it
compatible with other ones.

Signed-off-by: Bartosz Bilas <bartosz.bilas@hotmail.com>

mniestroj reviewed (approved):
LGTM

erwango commented:
@tbursztyka would you have time to review ?
